File: python/src/agent_framework/core/sub_agent_runner.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import AsyncIterator
from uuid import uuid4

from agent_framework.core.contracts import StreamEvent
from agent_framework.core.domain import AgentRunRequest, TerminalReason
from agent_framework.runtime import AgentRuntime


@dataclass
class SubAgentHandle:
    session_id: str
    parent_run_id: str
    run_id: str


class InProcessSubAgentRunner:
    def __init__(self, runtime: AgentRuntime) -> None:
        self._runtime = runtime
        self._completions: dict[str, TerminalReason] = {}

    async def run(self, request: dict) -> AsyncIterator[StreamEvent]:
        async for event in self._runtime.loop.run(
            AgentRunRequest(
                session_id=request["session_id"],
                input={"role": "user", "content": request["prompt"]},
            )
        ):
            yield event

    async def spawn(self, request: dict) -> SubAgentHandle:
        handle = SubAgentHandle(
            session_id=request["session_id"],
            parent_run_id=request["parent_run_id"],
            run_id=str(uuid4()),
        )

        async def _task() -> None:
            async for _event in self._runtime.loop.run(
                AgentRunRequest(
                    session_id=request["session_id"],
                    input={"role": "user", "content": request["prompt"]},
                )
            ):
                pass
            self._completions[handle.run_id] = {"kind": "completed"}

        asyncio.create_task(_task())
        return handle

    async def wait(self, handle: SubAgentHandle, timeout_ms: int = 60_000) -> TerminalReason:
        loop = asyncio.get_running_loop()
        deadline = loop.time() + timeout_ms / 1000
        while loop.time() < deadline:
            reason = self._completions.get(handle.run_id)
            if reason is not None:
                return reason
            await asyncio.sleep(0.01)
        return {"kind": "error", "message": "Sub-agent timed out"}

    async def cancel(self, handle: SubAgentHandle) -> None:
        self._completions[handle.run_id] = {"kind": "cancelled"}

File: typescript/packages/core/src/sub-agent-runner.ts
import { randomUUID } from 'node:crypto';
import type { AgentRuntime } from './runtime.js';
import type { StreamEvent, SubAgentRequest, SubAgentRunner } from './types/contracts.js';
import type { TerminalReason } from './types/domain.js';

export interface SubAgentHandle {
  sessionId: string;
  parentRunId: string;
  runId: string;
}

export class InProcessSubAgentRunner implements SubAgentRunner {
  private readonly completions = new Map<string, TerminalReason>();

  constructor(private readonly runtime: AgentRuntime) {}

  async *run(request: SubAgentRequest): AsyncGenerator<StreamEvent, TerminalReason, unknown> {
    const generator = this.runtime.loop.run({
      sessionId: request.sessionId,
      input: { role: 'user', content: request.prompt },
    });

    let result = await generator.next();
    while (!result.done) {
      yield result.value;
      result = await generator.next();
    }

    return result.value ?? { kind: 'completed' };
  }

  async spawn(request: SubAgentRequest): Promise<SubAgentHandle> {
    const handle: SubAgentHandle = {
      sessionId: request.sessionId,
      parentRunId: request.parentRunId,
      runId: randomUUID(),
    };

    void (async () => {
      const generator = this.runtime.loop.run({
        sessionId: request.sessionId,
        input: { role: 'user', content: request.prompt },
      });
      let result = await generator.next();
      while (!result.done) {
        result = await generator.next();
      }
      this.completions.set(handle.runId, result.value ?? { kind: 'completed' });
    })();

    return handle;
  }

  async wait(handle: SubAgentHandle, timeoutMs = 60_000): Promise<TerminalReason> {
    const started = Date.now();
    while (Date.now() - started < timeoutMs) {
      const reason = this.completions.get(handle.runId);
      if (reason) {
        return reason;
      }
      await new Promise((resolve) => setTimeout(resolve, 10));
    }
    return { kind: 'error', message: 'Sub-agent timed out' };
  }

  async cancel(handle: SubAgentHandle): Promise<void> {
    this.completions.set(handle.runId, { kind: 'cancelled' });
  }
}
